pinc/unicode.inc

```diff
@@ -170,3 +193,43 @@ function guess_string_encoding($text)
     # Give up and return ISO-8859-1
     return 'ISO-8859-1';
 }
+
+function make_title($codepoints)
+{
+    // make title from array of base codepoint and possible combining marks
+    $title = IntlChar::charName($codepoints[0]);
+    $count = count($codepoints);
+    if($count > 1)
+    {
+        $title .= " with " . IntlChar::charName($codepoints[1]);
+    }
+    for($index = 2; $index < $count; $index++)
+    {
+        $title .= " and " . IntlChar::charName($codepoints[$index]);
+    }
+    return $title;
+}
+
+function extended_to_hex($char)
+{
+    // find the components of extended character as hex strings
+    $codepoints = [];
+    preg_match_all("/./u", $char, $matches);
+    foreach($matches[0] as $match)
+    {
+        $codepoints[] = str_pad(dechex(IntlChar::ord($match)), 4, "0", STR_PAD_LEFT);
+    }
+    return implode(" ", $codepoints);
+}
+
+function extended_strlen($string)
```

Comment:
I can't work out what this does -- does it return the length of the string treating all combined characters as separate? Isn't that just the number of codepoints? If it does something different we need to rename it.

Reply:
This is a bit confusing. It returns the number of characters as it would appear if printed.
So the character Q̣̇ (with dot above and below) would have length 1 although it consists of 3 codepoints: U+0051 U+0323 U+0307 and 5 bytes: 51 cc a3 cc 87.
I've renamed it to combined_strlen.